teps/XXXX-step-workspaces.md

```diff
@@ -0,0 +1,254 @@
+      - name: my-workspace
+    ```
+
+The Sidecar `watch-workspace` will now have access to the `my-workspace` Workspace.
+
+### User Stories
+
+#### Story 1
+
+An author of the [`buildpacks-phases`](https://github.com/tektoncd/catalog/blob/master/task/buildpacks-phases/0.1/buildpacks-phases.yaml)
+Catalog task may want to rewrite the Task to reduce the possible blast radius of
+running untrusted images by limiting exposure of Docker credentials to only
+the Step which needs them to push images.
+
+In the buildpacks-phases Task there are 7 Steps and only 1 appears to need docker
+credentials. The are 6 other Steps that will currently receive creds-init docker
+credentials, run different images with different scripts and programs that could
+each be a vector to compromise those credentials.
+
+#### Story 2
+
+As a Task author I want to have a blessed method of sharing a Workspace between Steps and Sidecars
+so that I can pass configuration, messages, data, and logs between them.
+
+#### Story 3
+
+As a TaskRun and PipelineRun author I want to be able to trust that the Secret I provide
+via a Workspace Binding is only going to be available to code that is directly using it
+for its intended purpose.
+
+#### Story 4
+
+As a TaskRun and PipelineRun author I want to be able to quickly audit which
+containers, programs and scripts will have access to the credentials I am
+providing.
+
+### Risks and Mitigations
+
+## Design Details
+
+Add a new `WorkspaceUsage` struct to the `task_types.go` file:
+
+```go
+// WorkspaceUsage declares that a Step or Sidecar utilizes a Task's Workspace.
+type WorkspaceUsage struct {
+  // Name is the name of the Task's WorkspaceDeclaration that this Step or Sidecar is utilizing. It is required.
+  Name  string `json:"name"`
+}
+```
+
+Update the `Step` struct to include a slice of `WorkspaceUsage`:
+
+```go
+type Step struct {
+  corev1.Container `json:",inline"`
+  Script string `json:"script,omitempty"`
+
+  // Workspaces is a list of workspaces that this Step declares it will use in some way. The presence
+  // of a Workspace in this list prevents other Steps from automatically receiving the Workspace -
+  // they must also explicitly opt-in to receiving it as well in their own workspaces list.
+  Workspaces []WorkspaceUsage `json:"workspaces,omitempty"`
+}
+```
+
+Update the `Sidecar` struct to include a slice of `WorkspaceUsage`:
+
+```go
+// Sidecar embeds the Container type, which allows it to include fields not
+// provided by Container.
+type Sidecar struct {
+  corev1.Container `json:",inline"`
+  Script string `json:"script,omitempty"`
+
+  // Workspaces is a list of workspaces that this Sidecar declares it will use in some way.
+  Workspaces []WorkspaceUsage `json:"workspaces,omitempty"`
+}
+```
+
+## Drawbacks
+
+- It takes the `Step` type further from being a "pure" k8s Container. However we've already made moves
+in this direction by introducing our own `Script` field to `Steps`.
+
+## Alternatives
```

Comment:
Another alternative which is just a syntax tweak is instead of specifying a top level workspace AND referencing it in the step like:

```yaml
    workspaces:
    - name: my-sensitive-workspace
    steps:
    - name: foo
      workspaces:
      - name: my-sensitive-workspace
```

In the case where you want to limit access to the workspace to just a step, you could have the step contain all the info, i.e. the following would imply that the task needs a workspace `my-sensitive-workspace`

```yaml
    steps:
    - name: foo
      workspaces:
      - name: my-sensitive-workspace
```

Downsides:
1. If you want multiple steps to use the same workspace (and other steps not to), you have to rely on the same volume being provided at runtime - or you allow this to be indicated by using the same workspace name, e.g.:

```yaml
    steps:
    - name: foo
      workspaces:
      - name: my-sensitive-workspace
    - name: step-that-also-needs-sensitive-workspace
      workspaces:
      - name: my-sensitive-workspace
```
2. This is bit more confusing for the TaskRun / PipelineTask author in that in the above proposal, you know what workspaces you need to provide by looking at the list of workspaces; in this version you'd also need to look at the individual steps as well

Upsides:
1. Might help with @vdemeester 's point about different mount paths? i.e. if we moved the entire workspace definition into the step

Reply:
Possible way to work around downside (2) would be to introduce a "steps" stanza into PipelineTasks and TaskRuns, e.g. instead of:

```yaml
    workspaces:
    - name: my-workspace
    sidecars:
    - name: watch-workspace
      workspaces:
      - name: my-workspace
```

have something like:

```yaml
    workspaces:
    - name: my-workspace
    sidecars:
    - name: watch-workspace
      workspaces:
      - name: my-workspace
    steps:
    - name: sensitive-workspace
      workspaces:
      - name: some-sensitive-workspace
```

Reply:
I've added this as another alternative. I think naming collisions introduce a source of ambiguity here. First, the one you documented wrt Steps declaring the same Workspace name. It does make intuitive sense to me that the `Task` would only expose a single `Workspace` declaration for `TaskRuns` / `PipelineRuns` to bind to, and that binding would be shared across the Steps that include it, so this isn't so much of a problem to my mind.

What about the following example:

```yaml
workspaces:
- name: my-sensitive-workspace
steps:
- name: foo
  workspaces:
  - name: my-sensitive-workspace
- name: bar
  workspaces:
  - name: my-sensitive-workspace
```

Here ^ we've declared a workspace at both the Task level and individual Steps, all with the same name. Is this a validation error or does it follow the behavior of this proposal?  Again, not a big deal, just another rule to decide on and code in.

Finally, another source of potential validation error, consider the following:

```yaml
steps:
- name: foo
  workspaces:
  - name: my-sensitive-workspace
- name: bar
  workspaces:
  - name: my-senistive-workspace
```

Here, the validator won't be able to catch the mis-spelling of `my-senistive-workspace`.  If they're pre-declared at the top of the Task spec we can catch that and reject on submission, tightening the author's write-apply-test loop.  Again, not a _major_ concern I don't think but worth bearing in mind.

> Possible way to work around downside (2) would be to introduce a "steps" stanza into PipelineTasks and TaskRuns, e.g. instead of:
>     ```yaml
>     workspaces:
>     - name: my-workspace
>     sidecars:
>     - name: watch-workspace
>       workspaces:
>       - name: my-workspace
>     ```
> 
> have something like:
> 
>     ```yaml
>     workspaces:
>     - name: my-workspace
>     sidecars:
>     - name: watch-workspace
>       workspaces:
>       - name: my-workspace
>     steps:
>     - name: sensitive-workspace
>       workspaces:
>       - name: some-sensitive-workspace
>     ```

At the moment we don't expose either sidecars or steps for configuration in Runs and it feels to me like a change that doesn't serve a broad-enough purpose to warrant it.  However, if there _are_ good reasons to broaden the type changes this far I'm open to it.

Reply:
> Here ^ we've declared a workspace at both the Task level and individual Steps, all with the same name. Is this a validation error or does it follow the behavior of this proposal? Again, not a big deal, just another rule to decide on and code in.

```yaml
workspaces:
- name: my-sensitive-workspace
steps:
- name: foo
  workspaces:
  - name: my-sensitive-workspace
- name: bar
  workspaces:
  - name: my-sensitive-workspace
```

yeah that seems icky! I think this is a good point against this approach.

looking at that example (and the misspelling example) I like the version better where we declare the workspace in the workspaces list and reference it